README.rst
=======================
 `Drosophila eye map`_
=======================

.. _drosophila eye map: http://code.astraw.com/drosophila_eye_map

Overview
========

This software package concerns the eye-map of `Drosophila
melanogaster` made by Erich Buchner during his diplom thesis in
1971. These data were digitized from a high resolution scan of
Buchner's figure and converted to 3D and saved to the included file
``receptor_directions_buchner71.csv``.

Souce code repository
=====================

This source code is being hosted at
https://github.com/strawlab/drosophila_eye_map

Python package requirements
===========================

To use any of the included programs, you will need the Python_
language. For full functionality, this package depends on basemap_,
numpy_, scipy_, cgkit_ (1.x), matplotlib_, the `Python Imaging
Library`_, and, optionally, the Python VTK_ bindings. To simply use
the ``receptor_directions_buchner71.csv`` file, however, any program
which can open a CSV (comma separated values) file will work.

.. _Python: http://www.python.org/
.. _basemap: http://sourceforge.net/project/showfiles.php?group_id=80706&package_id=142792
.. _numpy: http://sourceforge.net/project/showfiles.php?group_id=1369&package_id=175103
.. _scipy: http://scipy.org/
.. _cgkit: http://sourceforge.net/project/showfiles.php?group_id=50475&package_id=44077&release_id=274256
.. _matplotlib: http://matplotlib.sourceforge.net/
.. _Python Imaging Library: http://www.pythonware.com/products/pil/
.. _VTK: http://www.vtk.org/

.. figure:: http://code.astraw.com/drosophila_eye_map/download/eye_map_small.gif
    :alt: Drosophila eye map
    :width: 400
    :height: 278
    :target: http://code.astraw.com/drosophila_eye_map/download/eye_map.gif

    The `Drosophila melanogaster` eye map of Buchner (1971) [#Buchner]_, as
    published in Heisenberg and Wolf (1984) [#Heisenberg]_ on page 11, Fig. 2.

.. figure:: http://code.astraw.com/drosophila_eye_map/download/interommatidial_distance_small.gif
   :alt: Drosophila eye map with interommatidial distance
   :width: 400
   :height: 300
   :target: http://code.astraw.com/drosophila_eye_map/download/interommatidial_distance.png

   Data of Buchner showing the eye map of `Drosophila melanogaster`
   eye map overlaid on a colormap showing interommatidial distance
   averaged over each ommatidium's nearest neighbors. (This image was
   generated using the
   ``make_buchner_interommatidial_distance_figure.py`` script included
   in the package.)

.. figure:: http://code.astraw.com/drosophila_eye_map/download/interommatidial_distance_ortho_small.gif
   :alt: Orthographic projection Drosophila eye map with interommatidial distance
   :width: 285
   :height: 300
   :target: http://code.astraw.com/drosophila_eye_map/download/interommatidial_distance_ortho.png

   Same data as the color stereographic projection above, but plotted
   on an orthographic projection for comparison with an `eyemap of
   Eristalis tenax`_.  (This image was generated using the
   ``make_buchner_interommatidial_distance_figure.py`` script included in
   the package.)

.. _eyemap of Eristalis tenax: http://jeb.biologists.org/cgi/content/full/209/21/4339/FIG1

Download
========

To download current and older versions of this package, go to the
`download page`_

.. _download page: http://code.astraw.com/drosophila_eye_map/download

History
=======

These programs and files have been extracted from Andrew Straw's `fsee
<https://github.com/strawlab/fsee>`_ software package for simulating
the visual world of Drosophila described in Dickson, Straw, and
Dickinson (2008) [#Dickson]_, part of the `GUF
<http://strawlab.org/2011/03/23/grand-unified-fly.html>`_ endeavor. At
the 2nd International Conference on Invertebrate Vision in Sweden,
2008, Andrew learned that a digitized form of the Drosophila eye map
would be useful to others, and so he created this package.

This is release 0.4, released Oct 19, 2008.

Contents of the package
=======================

In the top directory:

 * CHANGELOG.txt - list of changes since last release

 * LICENSE.txt - the (BSD) license

 * README.txt - this file

 * setup.py - script to install the software

 * upload_stuff.sh - script to release a package (only useful for
   maintainer)

 * drosophila_eye_map/ - subdirectory, see below

In the ``drosophila_eye_map`` subdirectory:

 * __init__.py - Empty file required for Python

 * inspect_weightmap.py - raphical program to inspect weightmap

 * make_buchner_interommatidial_distance_figure.py - Plot
   Buchner's data overlaid on a colormap showing mean interommatidial
   distance.

 * plot_receptors_vtk.py - Python script which is automatically
   inserted into the output of ``precompute_buchner71_optics.py``.

 * precompute_buchner71_optics.py - Python script used to take the
   output of ``trace_buchner_1971.py`` and convert it to a 3D
   coordinate system. Furthermore, a Gaussian spatial weighting map
   inspired by Neumann (2002) [#Neumann]_ is also implemented. These
   precomputed data are then saved for use by other programs as a file
   called ``precomputed_buchner71.py``.

 * receptor_directions_buchner71.csv - Comma separated value (CSV)
   file which indicates the directions of the ommaditial axes in 3D as
   vectors in a unit sphere. Output by
   ``precompte_buchner71_optics.py``. Note that this includes axes for
   both eyes (1398 ommatidia). To get the axes for a single eye, take
   the first or last 699 rows. The coordinate system is arranged so
   that +X is frontal (rostral), +Y is left, and +Z is dorsal.

 * trace_buchner_1971.py - Python script used to digitize the
   locations of the ommatidial axes on the stereographic projection of
   eye_map.gif__.

__ http://code.astraw.com/drosophila_eye_map/download/eye_map.gif

 * util.py - Utility routines used by
   ``precompute_buchner71_optics.py``.

License
=======

This software was written by Andrew Straw <andrew.straw@imp.ac.at>, is
copyright by the California Institute of Technology, and is licensed
under the BSD license. See the LICENSE.txt file for details.

Related software
================

`ArthroVision by the Invariant Corporation`_ is an insect optics
simulation package.

.. _ArthroVision by the Invariant Corporation: http://www.invariant-corp.com/arthrovision/

(See also the `History`_ section, above, for a description of Andrew
Straw's fsee software package.)

References
==========

.. [#Buchner] Buchner, E. (1971) `Dunkelanregung des stationaeren Flugs der Fruchtfliege Drosophila.` Dipl Thesis, Univ Tuebingen.

.. [#Heisenberg] Heisenberg, M. and Wolf, R., (1984) `Vision in Drosophila: Genetics Microbehavior (Studies of Brain Function).` Springer Verlag.

.. [#Dickson] Dickson, W.B., Straw, A.D., and Dickinson, M.H. (2008) "Integrative Model of Drosophila Flight." `AIAA Journal`, 46(9).  doi: `10.2514/1.29862`_

.. _10.2514/1.29862: http://dx.doi.org/10.2514/1.29862

.. [#Neumann] Neumann, T. (2002) "Modeling Insect Compound Eyes: Space-Variant Spherical Vision." `Biologically Motivated Computer Vision, Proceedings`. Springer Verlag. Vol 2525, pp. 360-367. doi: `10.1007/3-540-36181-2_36`_

.. _10.1007/3-540-36181-2_36: http://dx.doi.org/10.1007/3-540-36181-2_36


CHANGELOG.txt
Release 0.3
===========

* convert to cgkit 1.x

* update docs to point to new website location  in README.txt

* produce orthographic projection figure in
  make_buchner_interommatidial_distance_figure.py

* update references in README.txt

* add this file (CHANGELOG.txt)


drosophila_eye_map/make_buchner_interommatidial_distance_figure.py
# Copyright (c) 2005-2008, California Institute of Technology
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#
#     * Redistributions in binary form must reproduce the above
#       copyright notice, this list of conditions and the following
#       disclaimer in the documentation and/or other materials provided
#       with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# Author: Andrew D. Straw
from __future__ import division
import numpy as np
from scikits import delaunay as dlny

import drosophila_eye_map.precomputed_buchner71 as precomputed_buchner_1971
from drosophila_eye_map.util import get_mean_interommatidial_distance, xyz2lonlat
from mpl_toolkits.basemap import Basemap # basemap > 0.9.9.1

rdirs = precomputed_buchner_1971.receptor_dirs
rdir_slicer = precomputed_buchner_1971.receptor_dir_slicer
triangles = precomputed_buchner_1971.triangles

dists = np.array(get_mean_interommatidial_distance( rdirs[rdir_slicer['left']], triangles ))
R2D = 180.0/np.pi
dists = dists*R2D

lon_lats = [xyz2lonlat(*rdir) for rdir in rdirs[rdir_slicer['left']]]

stere = Basemap(projection='stere',
                resolution=None,
                lat_ts = 0.0,
                lat_0 = 0,
                lon_0 = 90,
                llcrnrlon = -45,
                urcrnrlon = -135,
                llcrnrlat= -30,
                urcrnrlat = 30,
                )

def do_projection( proj, lon_lats, dists, xres = 120, yres = 100 ):
    xys = np.array([proj( lon, lat ) for (lon,lat) in lon_lats ])

    x = xys[:,0]
    y = xys[:,1]

    good = x < 1e29 # basemap seems to set bad values to 1e30
    x=x[good]
    y=y[good]
    dists=dists[good]

    tri = dlny.Triangulation(x, y)
    interp = tri.nn_interpolator(dists)

    X,Y = np.mgrid[ min(y):max(y):yres*1j, min(x):max(x):xres*1j]
    vals = interp[ min(y):max(y):yres*1j, min(x):max(x):xres*1j]
    Z = np.ma.masked_array(vals,mask=np.isnan(vals))
    return x,y,X,Y,Z

x,y,X,Y,Z = do_projection(stere,lon_lats,dists)

import matplotlib
rcParams = matplotlib.rcParams
rcParams['font.size'] = 10
rcParams['font.family'] = 'serif'
#rcParams['font.serif'] = 'Times'
#rcParams['font.sans-serif'] = 'Arial'
if 0:
    from matplotlib import verbose
    verbose.level = 'debug-annoying'

import matplotlib.pyplot as plt
import matplotlib.cm

# pcolor figure -- stereographic projection
fig3 = plt.figure(3)
ax = plt.subplot(1,1,1)
plt.pcolor(Y,X,Z,shading='flat',cmap=matplotlib.cm.jet_r)

ax.plot(x,y,'wo',ms=4.0)

ax.text( 0.5,0.99, 'dorsal',
         horizontalalignment='center',
         verticalalignment='top',
         transform=ax.transAxes)
ax.text( 0.01,0.5, 'anterior',
         horizontalalignment='left',
         transform=ax.transAxes)
ax.text( 0.99,0.5, 'posterior',
         horizontalalignment='right',
         transform=ax.transAxes)

# draw parallels and meridians.
delat = 20.
circles = np.arange(0.,90.,delat).tolist()+\
          np.arange(-delat,-90,-delat).tolist()
#biw.stere.drawparallels(circles,ax=ax)
stere.drawparallels(circles,ax=ax)

delon = 45.
meridians = np.arange(-180,180,delon)
stere.drawmeridians(meridians,ax=ax)
cbar = plt.colorbar()
cbar.ax.set_ylabel('mean inter-ommatidial distance (deg)')

save_fname = 'interommatidial_distance.png'
print 'saving',save_fname
fig3.savefig(save_fname)
print 'OK'



# contour figure -- stereographic projection
fig1 = plt.figure(1)
ax = plt.subplot(1,1,1)
CS = plt.contour(Y,X,Z,
                   #levels=np.linspace(4,16,25),
                   levels=np.linspace(4,16,13),
                   colors='k',
                   )
plt.clabel(CS,
           fmt='%.1f',
           colors='k')

ax.plot(x,y,'wo',ms=4.0)

ax.text( 0.5,0.99, 'dorsal',
         horizontalalignment='center',
         verticalalignment='top',
         transform=ax.transAxes)
ax.text( 0.01,0.5, 'anterior',
         horizontalalignment='left',
         transform=ax.transAxes)
ax.text( 0.99,0.5, 'posterior',
         horizontalalignment='right',
         transform=ax.transAxes)

# draw parallels and meridians.
delat = 20.
circles = np.arange(0.,90.,delat).tolist()+\
          np.arange(-delat,-90,-delat).tolist()
stere.drawparallels(circles,ax=ax)

delon = 45.
meridians = np.arange(-180,180,delon)
stere.drawmeridians(meridians,
                    ax=ax)


# contour figure -- orthographic projection
fig2 = plt.figure(2)

# Match projection of
# http://jeb.biologists.org/cgi/content/full/209/21/4339/FIG1
ortho = Basemap(projection='ortho',
                lat_0=10,
                lon_0=20,
                resolution=None,
                )

x,y,X,Y,Z = do_projection(ortho,lon_lats,dists,xres=500,yres=500)

ax = plt.subplot(1,1,1)
CS = plt.contour(Y,X,Z,
                 #levels=np.linspace(4,16,25),
                 levels=np.arange(4,16.5,1.0),
                 colors='k',
                 linewidths=2.0,
                 )
plt.clabel(CS,fmt='%.1f',colors='k')

#ax.plot(x,y,'wo',ms=4.0)
ax.set_aspect('equal')

# draw parallels and meridians.
delat = 10.
circles = np.arange(0.,90.,delat).tolist()+\
          np.arange(-delat,-90,-delat).tolist()
circles = [c for c in circles if c!= 0.0]
ortho.drawparallels(circles,ax=ax)
ortho.drawparallels([0.0],linestyle='-',dashes=[],ax=ax)

delon = 10.
meridians = np.arange(-180,180,delon).tolist()
meridians = [m for m in meridians if m!=0.0]
ortho.drawmeridians(meridians,ax=ax)
ortho.drawmeridians([0.0],linestyle='-',dashes=[],ax=ax)
ortho.drawmapboundary(ax=ax)

# Text p. 12 of Heisenberg, M. and Wolf, R., (1984) implies this was a
# female.
#plt.title(unichr(2640)+' average inter-ommatidial angle (degrees)')
plt.title('female average inter-ommatidial angle (degrees)')

save_fname = 'interommatidial_distance_ortho.png'
print 'saving',save_fname
fig2.savefig(save_fname)
print 'OK'


plt.show()

